File: executioncontext/projectview/src/test/java/org/jetbrains/bsp/bazel/projectview/parser/splitter/ProjectViewRawSectionsTest.java
package org.jetbrains.bsp.bazel.projectview.parser.splitter;

import static org.assertj.core.api.Assertions.assertThat;

import io.vavr.collection.List;
import org.junit.jupiter.api.Test;

public class ProjectViewRawSectionsTest {

  @Test
  public void shouldReturnEmptyForEmptySections() {
    // given
    var sections = List.<ProjectViewRawSection>of();
    var projectViewRawSections = new ProjectViewRawSections(sections);

    // when
    var sectionWithName = projectViewRawSections.getLastSectionWithName("doesntexist");

    // then
    assertThat(sectionWithName).isEmpty();
  }

  @Test
  public void shouldReturnEmptyIfSectionDoesntExist() {
    // given
    var sections =
        List.of(
            new ProjectViewRawSection("name1", "body1"),
            new ProjectViewRawSection("name2", "body2"),
            new ProjectViewRawSection("name3", "body3"));
    var projectViewRawSections = new ProjectViewRawSections(sections);

    // when
    var sectionWithName = projectViewRawSections.getLastSectionWithName("doesntexist");

    // then
    assertThat(sectionWithName).isEmpty();
  }

  @Test
  public void shouldReturnOnlySectionWithName() {
    // given
    var sections =
        List.of(
            new ProjectViewRawSection("name1", "body1"),
            new ProjectViewRawSection("name2", "body2"),
            new ProjectViewRawSection("name3", "body3"));
    var projectViewRawSections = new ProjectViewRawSections(sections);

    // when
    var sectionWithName = projectViewRawSections.getLastSectionWithName("name1");

    // then
    var expectedSection = new ProjectViewRawSection("name1", "body1");
    assertThat(sectionWithName).containsExactly(expectedSection);
  }

  @Test
  public void shouldReturnLastSectionWithName() {
    // given
    var sections =
        List.of(
            new ProjectViewRawSection("name1", "body1.1"),
            new ProjectViewRawSection("name2", "body2"),
            new ProjectViewRawSection("name1", "body1.2"),
            new ProjectViewRawSection("name1", "body1.3"),
            new ProjectViewRawSection("name3", "body3"));
    var projectViewRawSections = new ProjectViewRawSections(sections);

    // when
    var sectionWithName = projectViewRawSections.getLastSectionWithName("name1");

    // then
    var expectedSection = new ProjectViewRawSection("name1", "body1.3");
    assertThat(sectionWithName).containsExactly(expectedSection);
  }

  @Test
  public void shouldReturnEmptyListForEmptySections() {
    // given
    var sections = List.<ProjectViewRawSection>of();
    var projectViewRawSections = new ProjectViewRawSections(sections);

    // when
    var sectionsWithName = projectViewRawSections.getAllWithName("doesntexist");

    // then
    assertThat(sectionsWithName).isEmpty();
  }

  @Test
  public void shouldReturnEmptyListIfSectionDoesntExist() {
    // given
    var sections =
        List.of(
            new ProjectViewRawSection("name1", "body1"),
            new ProjectViewRawSection("name2", "body2"),
            new ProjectViewRawSection("name3", "body3"));
    var projectViewRawSections = new ProjectViewRawSections(sections);

    // when
    var sectionsWithName = projectViewRawSections.getAllWithName("doesntexist");

    // then
    assertThat(sectionsWithName).isEmpty();
  }

  @Test
  public void shouldReturnAllSectionsWithName() {
    // given
    var sections =
        List.of(
            new ProjectViewRawSection("name1", "body1.1"),
            new ProjectViewRawSection("name2", "body2"),
            new ProjectViewRawSection("name1", "body1.2"),
            new ProjectViewRawSection("name1", "body1.3"),
            new ProjectViewRawSection("name3", "body3"));
    var projectViewRawSections = new ProjectViewRawSections(sections);

    // when
    var sectionsWithName = projectViewRawSections.getAllWithName("name1");

    // then
    var expectedSections =
        List.of(
            new ProjectViewRawSection("name1", "body1.1"),
            new ProjectViewRawSection("name1", "body1.3"),
            new ProjectViewRawSection("name1", "body1.2"));
    assertThat(sectionsWithName).containsExactlyInAnyOrderElementsOf(expectedSections);
  }
}

File: executioncontext/projectview/src/test/kotlin/org/jetbrains/bsp/bazel/projectview/parser/splitter/ProjectViewRawSectionsTest.kt
package org.jetbrains.bsp.bazel.projectview.parser.splitter

import io.kotest.matchers.collections.shouldContainExactlyInAnyOrder
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class ProjectViewRawSectionsTest {

    @Nested
    @DisplayName("fun getLastSectionWithName(sectionName: String): ProjectViewRawSection? tests")
    inner class GetLastSectionWithNameTest {

        @Test
        fun `should return null for empty sections`() {
            // given
            val sections = emptyList<ProjectViewRawSection>()
            val projectViewRawSections = ProjectViewRawSections(sections)

            // when
            val sectionWithName = projectViewRawSections.getLastSectionWithName("doesntexist")

            // then
            sectionWithName shouldBe null
        }

        @Test
        fun `should return null if section doesnt exist`() {
            // given
            val sections = listOf(
                ProjectViewRawSection("name1", "body1"),
                ProjectViewRawSection("name2", "body2"),
                ProjectViewRawSection("name3", "body3")
            )
            val projectViewRawSections = ProjectViewRawSections(sections)

            // when
            val sectionWithName = projectViewRawSections.getLastSectionWithName("doesntexist")

            // then
            sectionWithName shouldBe null
        }

        @Test
        fun `should return only section with name`() {
            // given
            val sections = listOf(
                ProjectViewRawSection("name1", "body1"),
                ProjectViewRawSection("name2", "body2"),
                ProjectViewRawSection("name3", "body3")
            )
            val projectViewRawSections = ProjectViewRawSections(sections)

            // when
            val sectionWithName = projectViewRawSections.getLastSectionWithName("name1")

            // then
            val expectedSection = ProjectViewRawSection("name1", "body1")
            sectionWithName shouldBe expectedSection
        }

        @Test
        fun `should return last section with name`() {
            // given
            val sections = listOf(
                ProjectViewRawSection("name1", "body1.1"),
                ProjectViewRawSection("name2", "body2"),
                ProjectViewRawSection("name1", "body1.2"),
                ProjectViewRawSection("name1", "body1.3"),
                ProjectViewRawSection("name3", "body3")
            )
            val projectViewRawSections = ProjectViewRawSections(sections)

            // when
            val sectionWithName = projectViewRawSections.getLastSectionWithName("name1")

            // then
            val expectedSection = ProjectViewRawSection("name1", "body1.3")
            sectionWithName shouldBe expectedSection
        }
    }

    @Nested
    @DisplayName("fun getAllWithName(sectionName: String): List<ProjectViewRawSection> tests")
    inner class GetAllWithNameTest {

        @Test
        fun `should return empty list for empty sections`() {
            // given
            val sections = emptyList<ProjectViewRawSection>()
            val projectViewRawSections = ProjectViewRawSections(sections)

            // when
            val sectionsWithName = projectViewRawSections.getAllWithName("doesntexist")

            // then
            sectionsWithName shouldBe emptyList()
        }

        @Test
        fun shouldReturnEmptyListIfSectionDoesntExist() {
            // given
            val sections = listOf(
                ProjectViewRawSection("name1", "body1"),
                ProjectViewRawSection("name2", "body2"),
                ProjectViewRawSection("name3", "body3")
            )
            val projectViewRawSections = ProjectViewRawSections(sections)

            // when
            val sectionsWithName = projectViewRawSections.getAllWithName("doesntexist")

            // then
            sectionsWithName shouldBe emptyList()
        }

        @Test
        fun shouldReturnAllSectionsWithName() {
            // given
            val sections = listOf(
                ProjectViewRawSection("name1", "body1.1"),
                ProjectViewRawSection("name2", "body2"),
                ProjectViewRawSection("name1", "body1.2"),
                ProjectViewRawSection("name1", "body1.3"),
                ProjectViewRawSection("name3", "body3")
            )
            val projectViewRawSections = ProjectViewRawSections(sections)

            // when
            val sectionsWithName = projectViewRawSections.getAllWithName("name1")

            // then
            val expectedSections = listOf(
                ProjectViewRawSection("name1", "body1.1"),
                ProjectViewRawSection("name1", "body1.3"),
                ProjectViewRawSection("name1", "body1.2")
            )
            sectionsWithName shouldContainExactlyInAnyOrder expectedSections
        }
    }
}
